### PR TITLE
Manual edits

### PR DESCRIPTION
**Pull request by jrscott**
edits to contributing section, git workflow changes

### CODE REVIEW

File: doc/contributing/contributing.rst [hunk 1]
```diff
@@ -31,49 +31,55 @@ To contribute to the source code of the model you will need to fork the reposito
...
-3. This step is optional, but highly recommended. It will ensure that your GitHub repo stays up to date with the main repository. Checkout the master branch and sync everywhere:
-
-::
-
-   % git checkout master
-   % git pull upstream master
-   % git push origin master
-   
-4. Make a new branch from ``upstream/master`` (name it something appropriate, here we assume it is a bugfix) and make edits on this branch:
+**4.** Make a new branch from ``upstream/master`` (name it something appropriate, here we assume it is a bugfix) and make edits on this branch:
```
Comment (edoddridge) [inline doc/contributing/contributing.rst]: The new branch is now called `newfeature` not `bugfix` -> should probably change this text to say "we assume you are adding a new feature"

**Comment by edoddridge** [comment]
@jklymak, @rabernat, and @jahn this is (potentially) the final version of the suggested workflow and contributors' guide to using git. It'd be great to have your opinions on it.

**Comment by jklymak** [comment]
I took a quick skim. Looks good.

The only issues I see going forward are 

  1. your `tags` file.  Matplotlib deals with this by having a `next-whats-new` directory where the user adds their whats-new entry as a `next-whats-new/jmk-new-feature-2018-01-16.rst`.  Then they don't have to rebase/merge a guaranteed conflicting file.  At release time, these are all concatenated (by hand or script, not sure) and given a new heading (Whats New 2.1.2). 
  2. How are you going to handle releases?  I appreciate you want to have some fluidity, and that seems fine, but do you really want all users working off Master?  It kind of sounds like it, and that seems tough to maintain.  

Anyway, I'm sure you already thought about both of these issues.  Otherwise, looks good!  Lots of work; thanks a lot for doing it!

File: doc/contributing/contributing.rst [hunk 2]
```diff
@@ -31,49 +31,55 @@ To contribute to the source code of the model you will need to fork the reposito
 Quickstart Guide
 ----------------
 
-1. Fork the project on GitHub (using the fork button) and create a local clone, if you haven’t already:
+**1.** Fork the project on GitHub (using the fork button).
+
+**2.** Create a local clone (we strongly suggest keeping a separate repository for development work):
```
Comment (rabernat) [inline doc/contributing/contributing.rst:36]: It seems unnecessary and confusing to say "we strongly suggest keeping a separate repository for development work". This guide is for contributors. There is literally no other way to contribute.

File: doc/contributing/contributing.rst [hunk 3]
```diff
@@ -31,49 +31,55 @@ To contribute to the source code of the model you will need to fork the reposito
...
 
-2. Move into your local clone directory (cd MITgcm) and and set up a remote that points to the original:
+**3.** Move into your local clone directory (cd MITgcm) and and set up a remote that points to the original:
 
 ::
 
     % git remote add upstream https://github.com/altMITgcm/MITgcm.git
 
-3. This step is optional, but highly recommended. It will ensure that your GitHub repo stays up to date with the main repository. Checkout the master branch and sync everywhere:
+**4.** Make a new branch from ``upstream/master`` (name it something appropriate, here we assume it is a new feature) and make edits on this branch:
```
Comment (rabernat) [inline doc/contributing/contributing.rst]: "here we name the branch ``newfeature``"... this would connect more directly to the command below

File: doc/contributing/contributing.rst [hunk 4]
```diff
@@ -31,49 +31,55 @@ To contribute to the source code of the model you will need to fork the reposito
...
-   % git checkout master
-   % git pull upstream master
-   % git push origin master
-   
-4. Make a new branch from ``upstream/master`` (name it something appropriate, here we assume it is a bugfix) and make edits on this branch:
+   % git fetch upstream
+   % git checkout -b newfeature upstream/master
 
-::
+**5.** When edits are done, do all git add’s and git commit’s. In the commit, make a succinct (<70 char) summary, followed by a blank line and a longer description. Reference any outstanding issues addressed using the syntax ``#ISSUE_NUMBER``.
```
Comment (rabernat) [inline doc/contributing/contributing.rst]: I don't think a longer description belongs in the commit history itself. The longer description can come when the user makes the actual PR via github. It is common for one PR to have many commits ([example](https://github.com/pydata/xarray/pull/1796/commits)), which are then squashed at merge time. Each individual commit just needs the succinct summary.

File: doc/contributing/contributing.rst [hunk 5]
```diff
@@ -31,49 +31,55 @@ To contribute to the source code of the model you will need to fork the reposito
...
+**7.** On GitHub, go to your fork and hit the pull request (PR) button, and wait for the MITgcm head developers to review your proposed changes. You may get additional questions or requests before your changes are accepted into the primary MITgcm source code.
+
+If you want to update your code branch before submitting (or any point in development), follow the following recipe. It will ensure that your GitHub repo stays up to date with the main repository. Note again that your edits should always be to a development branch (here, “newfeature”), not the master branch.
 
 ::
 
-    % git push -u origin bugfix
+    % git checkout master
+    % git pull upstream master
+    % git push origin master
```
Comment (rabernat) [inline doc/contributing/contributing.rst:79]: why is this necessary?

File: doc/contributing/contributing.rst [hunk 6]
```diff
@@ -31,49 +31,55 @@ To contribute to the source code of the model you will need to fork the reposito
...
 
-5. When edits are done, do all git add’s and git commit’s. In the commit, make a succinct (<70 char) summary, followed by a blank line and a longer description. Reference any outstanding issues addressed using the syntax ``#ISSUE_NUMBER``.
+::
 
+    % git push -u origin newfeature
 
-6. Push the edited branch to the origin remote (i.e. your fork) on GitHub:
+**7.** On GitHub, go to your fork and hit the pull request (PR) button, and wait for the MITgcm head developers to review your proposed changes. You may get additional questions or requests before your changes are accepted into the primary MITgcm source code.
+
+If you want to update your code branch before submitting (or any point in development), follow the following recipe. It will ensure that your GitHub repo stays up to date with the main repository. Note again that your edits should always be to a development branch (here, “newfeature”), not the master branch.
```
Comment (rabernat) [inline doc/contributing/contributing.rst]: Clarify what is meant by "submitting"? You mean submitting a PR? If so, say so.

File: doc/contributing/contributing.rst [hunk 7]
```diff
@@ -31,49 +31,55 @@ To contribute to the source code of the model you will need to fork the reposito
...
-   % git checkout -b bugfix upstream/master
+**6.** Push the edited branch to the origin remote (i.e. your fork) on GitHub:
 
-5. When edits are done, do all git add’s and git commit’s. In the commit, make a succinct (<70 char) summary, followed by a blank line and a longer description. Reference any outstanding issues addressed using the syntax ``#ISSUE_NUMBER``.
+::
 
+    % git push -u origin newfeature
 
-6. Push the edited branch to the origin remote (i.e. your fork) on GitHub:
+**7.** On GitHub, go to your fork and hit the pull request (PR) button, and wait for the MITgcm head developers to review your proposed changes. You may get additional questions or requests before your changes are accepted into the primary MITgcm source code.
```
Comment (rabernat) [inline doc/contributing/contributing.rst]: Should we give explicit instructions on how to update a PR in response to a request for changes? I think so.

- Possibly update your feature branch if master has changed significantly (see instructions below) 
- Edit your `newfeature` branch locally
- `git push origin newfeature`
- Your PR will automatically be updated

**Comment by rabernat** [comment]
Regarding the "tags" / "whats-new" question: in xarray we just have a single master file in the docs called whats-new.rst. Every pull request needs to include an edit to this file describing the changes. We never have any issues merging this file.

**Comment by rabernat** [comment]
@jklymak is spot on about releases. I strongly encourage the use of github's release mechanism:
https://help.github.com/articles/creating-releases/

The notion of releases syncs up roughly with our concept of checkpoints. Whenever we think it's time for checkpoint, we create a release. Readthedocs will also automatically manage the releases and build each version of the documentation appropriately. So users can be sure to have the correct documentation for the release they are actually using.

File: doc/contributing/contributing.rst (hunk 2 above)
Comment (edoddridge) [reply doc/contributing/contributing.rst:36]: This is intended to steer people away from using the same set of local files for both running the model for science and developing the model. 

We could rephrase to "...keeping a separate set of local files for development work."

File: doc/contributing/contributing.rst (hunk 3 above)
Comment (edoddridge) [reply doc/contributing/contributing.rst]: OK - update is on the way.

File: doc/contributing/contributing.rst (hunk 4 above)
Comment (edoddridge) [reply doc/contributing/contributing.rst]: OK, how about we rephrase this to 
"In the commit message, make a succinct (<70 char) summary of your changes. If you need more space to describe your changes, you can leave a blank line and type a longer description, or break your commit into multiple smaller commits."

File: doc/contributing/contributing.rst (hunk 7 above)
Comment (edoddridge) [reply doc/contributing/contributing.rst]: We do give more detailed instructions, but not in the quickstart guide, since we assume people reading that are somewhat familiar with git and GitHub. Step 7 of the longer version includes the following:
"While the PR remains open, you can go back to step #5 and make additional edits, git adds, git commits, and then redo step #6; such changes will be added to the PR (and maintainers re-notified), no need to redo step #7."

File: doc/contributing/contributing.rst (hunk 6 above)
Comment (edoddridge) [reply doc/contributing/contributing.rst]: Good thought. @jrscott is adding that.

File: doc/contributing/contributing.rst (hunk 5 above)
Comment (edoddridge) [reply doc/contributing/contributing.rst:79]: Our git guinea pigs told us that they were confused when their fork didn't stay up to date with the main repo. These commands ensure their fork stays up to date. But, yes, they're not really necessary.

Comment (rabernat) [reply doc/contributing/contributing.rst:79]: That's a fair point. But I think that telling people to `push origin master` here creates false impression that it is necessary part of the process, which ultimately will cause greater confusion.

Once the fork has been made, origin/master never needs to be touched again. You merge directly from upstream/master to the local master.

Comment (jklymak) [reply doc/contributing/contributing.rst:79]: Your git guinea pigs and everyone else should be encouraged to *not* have a `master` branch in their development distro.  Then they won't be confused.  

They should not be checking out `master` they should be checking out `placeholder` or something anodyne.  They should periodically 

```
git fetch upstream master
git checkout placeholder
git reset --hard upstream/master
git checkout -b mypristinenewbranchoffmaster
```

This makes it crystal clear that `origin` is just a place for my changes, and doesn't mean anything except as a reference for PRs.  If you dont' do this, people are going to think `origin/master` is magically `upstream/master` somehow.

Comment (jklymak) [reply doc/contributing/contributing.rst:79]: http://matthew-brett.github.io/pydagogue/gh_delete_master.html

**Comment by edoddridge** [comment]
Thanks @jklymak and @rabernat! 

Regarding `doc/tag-index` vs some other way of recording what's new: `genmake2` parses `doc/tag-index` to find the latest checkpoint and puts that in the standard output. We can change that behaviour, but one of our aims is to avoid code changes as part of the transition to git.

As for releases, our (perhaps unstated?) intention was to use them like checkpoints are currently used, just as @rabernat suggested.

File: doc/contributing/contributing.rst (hunk 5 above)
Comment (edoddridge) [reply doc/contributing/contributing.rst:79]: I agree with you both, especially @rabernat's point about people potentially thinking `git push origin master` is a necessary part of the process; `origin/master` is not a branch that should ever be used. But, we asked people to tell us what they found difficult/off putting/strange about the new workflow and we got feedback from multiple people saying they were confused about why their repository wasn't staying up to date. Inserting a (somewhat superfluous) `git push origin master` is an easy way to avoid that confusion.

As much as I like the idea of telling people to remove `master` from their fork, I don't think those instructions are beginner friendly. Perhaps it would be better to think of this section of the manual as a "git survival guide for the uninitiated", rather than a description of the best possible workflow.

If you two are dead against having `git push origin master` in there, then I guess we'll have to keep working on this section.

Comment (rabernat) [reply doc/contributing/contributing.rst:79]: It's fine as is. We are splitting hairs.

Better to just get something up there and then iterate in the future if it seems necessary.

File: doc/contributing/contributing.rst [hunk 8]
```diff
@@ -167,7 +167,7 @@ You will make edits on this new branch, to keep these new edits completely separ
...
+    - use ``git commit`` to commit the files. This is the first step in bundling a collection of files together to be sent off to the MITgcm maintainers. When you enter this command, an editor window will pop up. On the top line, type a succinct (<70 character) summary of what these changes accomplished. If your commit is non-trivial and addition explanation is required, leave a blank line and then type a longer description of why the action in this commit was appropriate etc. It is good practice to link with known issues using the syntax ``#ISSUE_NUMBER`` in either the summary line or detailed comment. Note that all the changes do not have to be handled in a single commit (i.e. you can git add some files, do a commit, than continue anew by adding different files, do another commit etc.); the ``git commit`` command itself does not (yet) submit anything to maintainers.
```
Comment (edoddridge) [inline doc/contributing/contributing.rst]: There's a typo here: "If your commit is non-trivial and addition explanation..." Should be "additional"

**Comment by edoddridge** [comment]
LGTM

**Comment by jahn** [comment]
I put the table example back for all builders, converted to just a set of simple tables, so it doesn't break the latex builder.  LGTM
